lib/Sema/CSDiag.cpp

```diff
@@ -6223,11 +6223,47 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
   if (diagnoseCallContextualConversionErrors(callExpr, CS->getContextualType()))
     return true;
 
-  // Type check the function subexpression to resolve a type for it if
-  // possible.
-  auto fnExpr = typeCheckChildIndependently(callExpr->getFn());
-  if (!fnExpr)
+  auto originalFnType = CS->getType(callExpr->getFn());
+
+  // Check if there is a structural problem in the function expression
+  // itself, by type-checking it allowing unresolved variables. If
+  // that is going to produce a functional type with unresolved result
+  // let's not re-typecheck the function expression, because it might
+  // produce unrelated diagnostics due to lack of contextual information.
+  auto shouldTypeCheckFunction = [&](Expr *fnExpr) -> bool {
```

Comment:
meta comment: `visitApplyExpr` is quite large. Rather than inlining lambdas like this, it probably makes sense to break them out into separate functions. It's already really hard to follow a lot of code due to the interspersing of specific decisions about how to diagnose problems and the support code for those decisions.